### ISSUE TITLE
Enhancement Requests for Haxfootball Script

### ISSUE THREAD

**Issue by anarcer** [haxball/haxball-issues#525]
Thanks folks for spending the time to create APIs with the headless host, they've been great! I've spent the last few weeks creating a bot that can support custom game style/map Haxfootball (American football) in my spare time. It's been working really well, and something that's helped improve the game immensely.

A few enhancements that would even make the bot more useful!
1. Ability to set player + ball positions. This would allow us to reset the ball to the most appropriate yard when a player gains yards in the game.

2.  Ability to detect if the player is currently holding X (flash). To signal a run play in the game, we typically want the Quarterback to hold X when touching the running back. With the current APIs, I can obviously detect player collision using point distance formulas, however I can't detect whether the player is currently holding X or not. Would be nice to get this added as well.

Side note - I can test and confirm. However, the array returned by room.playerlist shouldn't change unless a player leaves/joins or gets moved to a different team right? I'm trying to remove performance debt in the current script, and don't want to unnecessarily ping the room for a new player list when I don't think that array would change. (Currently i'm pinging at each tick, which I don't think is needed).

**Comment by anarcer** [comment]
Nevermind regarding #1, saw it was requested in #425 already.

**Comment by iamrenan** [comment]
Hi anarcer,

Just a fellow coder here too, hoping these API functions you mentioned get
added someday, as they'd be very useful!

But regarding your side note, yes, the getPlayerList() return all the
players in the room as PlayerObjects - As seen in the docs, the
PlayerObject has "id, name, team, admin, and position" as fields, so if any
of these values change (or the object itself), calling getPlayerList() will
return a different array with the changed values. For most applications
though, you only want to call it when you call onPlayerJoin(),
onPlayerLeave() and/or onPlayerTeamChange(), as you're currently doing :)

Em sáb, 15 de set de 2018 às 12:03, anarcer <notifications@github.com>
escreveu:

**Comment by anarcer** [comment]
Yep, thanks! I realized yesterday in order for me to get updated position of the players, I will need to get an updated playerObject from the room at every tick. I wasn't thinking through it, and for some reason assumed the object I created would automatically update (silly me).

I did find out that the index of the array in getPlayerList only changes when players are moved, or a player joins/leaves. That should help a little bit.

Thanks for the comment!

**Comment by iamrenan** [comment]
Oh, yeah, sorry, disregard what I said about "as you're currently doing" lol

But np! I didn't used onGameTick yet, but I believe using a setInterval
could be better performance-wise, as long as you set it to something higher
than 16ms (I think)

Em dom, 16 de set de 2018 11:37, anarcer <notifications@github.com>
escreveu:

**Comment by basro** [comment]
@anarcer: sorry, this feature didn't make it into this update. I'll consider it in the future.

**Comment by anarcer** [comment]
Basro, can you add the "Feature Request" tag to this? Just so it doesn't drop off the list.

As an update, we have a fairly active community now. The game play is unique in haxball, and completely dependent on the bot. (You really can't play american football without the bot, since bot does the tackle calculations, down/distance calculations, etc.) The bot isn't just used for chat functions or stats collection.

These three features would be really beneficial:
1. Ability to set the ball position
2. Ability to start/stop the clock (not as important as 1.)
3. Ability to set play positions.

Side Note: Would be cool to add a 'Team Chat' function instead of just PM chat. Currently my bot sends the chat to the rest of the team if someone starts their chat with a semicolon. So definitely not a rush, but figured it would be a good addition down the line.

**Comment by anarcer** [comment]
@basro it's been 6 months, and I was wondering if you have given any more thoughts about being able to set the ball (and potentially player) position through the bot? It would be such an insane improvement to our game play that I can't resist asking again.

**Comment by Dime219** [comment]
Hey @basro, as a member of the Haxfootball community, I can confirm that these features would greatly improve the game. I believe adding them to your next planned update would be a positive step forward for Haxfootball and other maps/modes with bot hosts. This is a friendly reminder as this post seems to have been lost in the mix :))
